### ISSUE TITLE
Add support for variable locking

### ISSUE THREAD

**Issue by muglug** [vimeo/psalm#103]
Once a variable is assigned by reference, we should support locking the pointers together (maybe by way of a special type).

```php
function takesInt(int $i) { }

$a = 5;
$b = &$a;
$a = "hello";
takesInt($b);
```

The last line should generate an error; currently it does not.

**Comment by muglug** [comment]
Not sure what we should do here though

```php
function takesInt(int $i) { }

$a = 5;
if (rand(0, 1)) {
    $b = 6;
} else {
    $b = &$a;
}
$a = "hello";
takesInt($b);
```

The type of `$a` at the end is `string`
The type of `$b` at the end is `int|typeof($a)` = `int|string`

**Comment by muglug** [comment]
Closing as out of scope
